File: code_combat/8_Cloudrip_Mountain/479-Double_Agent/doubleAgent.js
// Find the hidden number in the agent's message to escape.
// Count the number of trailing and leading whitespaces.

// This function returns the coordinates of the n-th passage.
function passagePosByNum(n) {
    return {x: 60, y: n * 12 + 8};
}

function onHear(event) {
    // The original message.
    var message = event.message;
    // Trim the message:
    var trimmed = message.trim();
    // The hidden number is the difference of lengths:
    var hidden = message.length - trimmed.length;
    // Use passagePosByNum to find the passage to enter:
    var pos = passagePosByNum(hidden);
    // Move the pet to the entrance of the passage:
    pet.moveXY(pos.x, pos.y);
    // Move the pet to the left edge of the map:
    pet.moveXY(2, pos.y);
}

pet.on("hear", onHear);

// The hero should follow the pet.
while(true) {
    hero.move(pet.pos);
}

File: code_combat/8_Cloudrip_Mountain/479-Double_Agent/double_agent.py
def passagePosByNum(n):
    return {"x": 60, "y": n * 12 + 8}


def onHear(event):
    message = event.message
    trimmed = message.strip()
    hidden = len(message) - len(trimmed)
    pos = passagePosByNum(hidden)
    pet.moveXY(pos.x, pos.y)
    pet.moveXY(2, pos.y)


pet.on("hear", onHear)

while True:
    hero.move(pet.pos)
